synapse/types.py

```diff
@@ -400,6 +400,37 @@ def f2(m: Match[bytes]) -> bytes:
     return username.decode("ascii")
 
 
+MU = TypeVar("MU", bound="MXCUri")
+
+
+@attr.s(frozen=True, slots=True, auto_attribs=True, repr=False)
+class MXCUri:
+    """Represents a URI that points to a media resource in matrix.
+
+    MXC's take the form 'mxc://server_name/media_id'.
+    """
+
+    server_name: str
+    media_id: str
+
+    @classmethod
+    def from_str(cls: Type[MU], mxc_uri_str: str) -> MU:
+        try:
+            server_name, media_id = mxc_uri_str.split("/")[-2:]
```

Comment:
This accepts any string containing at least one forward slash. For instance, the string `https://github.com/anoadragon453/nixos-config/` would yield `MXCUri('nixos-config', '')`.

Maybe a regex is better suited for this?

Also, is a `media_id` allowed to contain forward slashes?

Reply:
Or maybe we should use [`urlparse`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse)?

Reply:
Correct, but at least it will fully parse it and we can then do more assertions on it, instead of writing our own URL parsing logic.